Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,342 @@
+
+## Main statements
+
+* `t2Space_of_ProperSMul`: If a group `G` acts properly on a topological space `X`,
+  then the quotient space is Hausdorff (T2).
+* `t2Space_of_properSMul_of_t2Group`: If a T2 group acts properly on a topological space,
+  then this topological space is T2.
+
+## Notation
+
+
+
+## Implementation details
+
+
+
+## References
+
+* [F. Bar, *Quuxes*][bibkey]
+
+## Tags
+
+Foobars, barfoos
+-/
+
+
+
+
+/-- Additive version of proper action in the sense of Bourbaki:
+the map `G×X→ X×X` is a proper map `isProperMap`
+-/
+@[mk_iff]
+class ProperVAdd (G X : Type*) [TopologicalSpace G] [TopologicalSpace X] [AddGroup G]
+    [AddAction G X] : Prop where
+  /-- Additive version of proper action in the sense of Bourbaki:
+the map `G×X→ X×X` is a proper map `isProperMap`  -/
+  isProperMap_vadd_pair' : IsProperMap (fun gx ↦ ⟨gx.1 +ᵥ gx.2, gx.2⟩ : G × X → X × X)
+
+/-- Proper action in the sense of Bourbaki:
+the map `G×X→ X×X` is a proper map `isProperMap`
+-/
+@[to_additive existing, mk_iff]
+class ProperSMul (G X : Type*) [TopologicalSpace G] [TopologicalSpace X] [Group G]
+    [MulAction G X] : Prop where
+  isProperMap_smul_pair' : IsProperMap (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X)
+
+attribute [to_additive existing] properSMul_iff
+
+/-- By definition, if G acts properly on X
+the map `G×X→ X×X` is a proper map `isProperMap`
+-/
+@[to_additive]
+lemma isProperMap_smul_pair (G X : Type*) [Group G] [MulAction G X]
+    [TopologicalSpace G] [TopologicalSpace X] [ProperSMul G X] :
+    IsProperMap (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) :=
+  ProperSMul.isProperMap_smul_pair'
+
+open Filter Topology
+
+variable {G X Y Z W : Type*} [Group G] [MulAction G X] [MulAction G Y]
+variable [TopologicalSpace G] [TopologicalSpace X] [TopologicalSpace Y]
+variable [TopologicalSpace Z] [TopologicalSpace W]
+
+@[to_additive]
+instance continuousSmul_of_properSMul [ProperSMul G X] : ContinuousSMul G X where
+  continuous_smul := (isProperMap_smul_pair G X).continuous.fst
+
+/-- A group `G` acts properly on a topological space `X` if and only if for all ultrafilters`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `g • x₂ = x₁` and `𝒰.fst` converges to `g`. -/
+@[to_additive "A group acts `G` properly on a topological space `X` if and only if
+for all ultrafilters `𝒰` on `X`, if `𝒰` converges to `(x₁, x₂)`
+along the map `(g, x) ↦ (g • x, x)`, then there exists `g : G` such that `g • x₂ = x₁`
+and `𝒰.fst` converges to `g`."]
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto : ProperSMul G X ↔ ContinuousSMul G X
+    ∧ (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, g • x₂ = x₁ ∧ Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  constructor
+  · intro h
+    refine ⟨by infer_instance, fun 𝒰 x₁ x₂ h' ↦ ?_⟩
```

Comment:
This is a very small nitpick: instead of `by infer_instance`, you can directly use the term-mode equivalent to the tactic `infer_instance`, which is `inferInstance`.
```suggestion
    refine ⟨inferInstance, fun 𝒰 x₁ x₂ h' ↦ ?_⟩
```